New NixOS module (from #306)

pjones opened the pull request:
Introduce a new NixOS module and NixOS test.  This is a starting point
for replacing the current module with something a bit easier to use
and more flexible.

david-janssen commented:
Hello @pjones,

Thank you for your contribution. I would very much like to make the nix-plumbing for KMonad work better than it currently does. My issue is that I really like using Nix, but I do not understand it very well. I am going to wait a few days to see if @slotThe has any input about this pull-request. He is much more aware of the current state of KMonad than I am. If he okays it, we can merge.

Since you mention this is a 'starting point' I was wondering if you'd be happy to help me get the rest of the nix-plumbing working a bit better? Currently simply calling `nix-shell` in the top-level doesn't work, and calling `nix-build nix` will build a dev-environment with some GHC, but won't produce a binary.

Ideally I'd be able to run a `nix-shell` that gets me all of the dev-tools I need to hack on KMonad, and a `nix-build` to produce/install a binary that I can then run. However, since I first wrote (duct-taped together) the nix-code more 18 months ago, I've gotten a bit ill and have had trouble focusing. I'd like to focus on investing what available thinky-time I have into KMonad, and first getting nix 'working properly' could take me a long time, since it's always been a bit of a duct-tape operation.

Cheerio and all the best,

David

pjones commented:
@david-janssen I'd be happy to help with the Nix stuff.  I was actually planning on submitting another PR to fix the flake.  I was having some weird issues with it and ended up having to [write my own](https://github.com/pjones/tilde/blob/496b743a318dac29e6f68ab80be7f9cd7388be6e/pkgs/kmonad/flake.nix).  I also have an example flake file that contains a fully functional Haskell development environment [here](https://github.com/pjones/turtle/blob/01553ff215df1a8b9d5b9536f3a47a49faa6bb6b/flake.nix).  I could merge those two flakes into a replacement flake/default.nix file for KMonad.

As far as this PR goes, I think we need to offer a few module options so users can change all of the settings in the `defcfg` block.  Once that is done I think it would be ready to merge if everyone agrees with the approach.

david-janssen commented:
> I could merge those two flakes into a replacement flake/default.nix

That would be superlative :-)

> I think we need to offer a few module options so users can change all of the settings in the defcfg block. Once that is done I think it would be ready to merge if everyone agrees with the approach.

That sounds like it makes sense. Is there anything you would like me to do first, or are you happy to make these changes when time permits and I await your proposal?

pjones commented:
@david-janssen Should I make all of my Nix-related changes part of this PR?

david-janssen commented:
@pjones Either way is alright with me. We could merge the new nixos module first, and then move on to the other nix stuff, or we could make everything good in this PR and merge it in all at one. However you prefer.

A quick question: I have NixOS, but I've never used non builtin modules. Would I be correct in thinking that, to use this kmonad module I would:

```nix
  imports = [ ./path/to/kmonad-module.nix ];

  services.kmonad =
    { enable = true;
      other    = 5;
      stuff     = "abba"
      here    = 42; };
```

pjones commented:
> A quick question: I have NixOS, but I've never used non builtin modules. Would I be correct in thinking that, to use this kmonad module I would:

Yes, that's correct.

The actual path (obviously) depends on where/how you reference
KMonad.  This is much easier with Nix flakes because you can use the
flake input to reference the module, something like:

```nix
{
  description = "Example Nix Flake w/ KMonad";

  inputs = {
    nixpkgs.url = "...";
    kmonad.url = "...";
  };

  outputs = { nixpkgs, inputs, ...}: {
    nixosConfigurations.myhost = nixpkgs.lib.nixosSystem {
      system = "x86_64-linux";
      modules = [
        inputs.kmonad.nixosModules.kmonad
        ./path/to/nixos-configuration.nix
      ];
    };
  };
}
```

david-janssen commented:
That looks really interesting. I will read up on Nix flakes and figure out how to incorporate them into my personal config. Thank you for pointing me in this direction.

At the moment I can't see what I can contribute further to this PR, so unless you have any questions or comments, I am going to patiently await progress and see what bubbles up.

pjones commented:
I've added the remaining features to the NixOS module.  I've also added some additional NixOS tests that actually switch the simulated keyboard to the numbers layer and ensure numbers are produced.

I think we should merge this PR and then I'll open another one with the remaining Nix changes.

david-janssen commented:
Merged. And also: I had a chance to try out nix flakes on a little toy project yesterday. It was so much easier than any nix haskell solution I've tried before. Thanks for the tip, I look forward to setting up the same infrastructure for KMonad :)

david-janssen commented on nix/module-base.nix:
```diff
@@ -1,127 +1,142 @@
-      description = ''
-        The kmonad package.
-      '';
+      example = "pkgs.haskellPacakges.kmonad";
```
Hey Andrew, 

Thanks for the headsup, however I think that the `nix/module-base.nix` file got moved in #549. However, I found the same typo in that pull-request on line 121, so... Well spotted!

Fixed :)
